[Staging] (DO NOT MERGE UNTIL 7/26) Microsoft Teams Get channel messages API's

hamenon-ms opened the pull request:
Pull request for updating Graph Documentation with the GET channel messages and GET channel message replies API's. At a high level the changes are as follows

- Updating the chatmessage resource
- Adding the chatattachment, chatmention and chatreaction resources
- Adding the new API's for List all channel messages, get a channel message, list all replies to a message and get a reply to a channel message

FYI @nkramer

nkramer commented on api-reference/beta/api/channel_message_get.md:
```diff
@@ -0,0 +1,193 @@
+|---------|-------------|
+|Delegated (work or school account)|Group.Read.All,Group.ReadWrite.All|
+|Delegated (personal Microsoft account)|Not supported|
+|Application|Group.Read.All, Group.ReadWrite.All|
```
The initial implementation won't support application permissions.

Sunit41557077 replied:
<?xml version="1.0" encoding="UTF-8"?>
<xsd:schema xmlns:xsd="http://www.w3.org/2001/XMLSchema"
    xmlns="http://cve.mitre.org/cve/downloads/1.0"
    targetNamespace="http://cve.mitre.org/cve/downloads/1.0"
    elementFormDefault="qualified" attributeFormDefault="unqualified" version="1.0">

    <!--  **************************************************************************************** -->
    <!--  Changelog:
                1.0 - Initial version
    -->
    <!--  **************************************************************************************** -->
    <xsd:annotation>
        <xsd:documentation xml:lang="en"> Simple schema that defines the format of the CVE List
            provided by MITRE </xsd:documentation>
    </xsd:annotation>

    <!--  **************************************************************************************** -->
    <!--  Start Item Element Definition -->
    <!--  **************************************************************************************** -->
    <xsd:element name="cve">
        <xsd:annotation>
            <xsd:documentation xml:lang="en"> cve is the top level element of the CVE List provided
                by MITRE. It represents holds all CVE Items. </xsd:documentation>
        </xsd:annotation>
        <xsd:complexType>
            <xsd:sequence>
                <xsd:element name="item" type="ItemType" minOccurs="1" maxOccurs="unbounded"/>
            </xsd:sequence>
            <xsd:attribute name="schemaVersion" type="xsd:token" use="optional"/>
        </xsd:complexType>
    </xsd:element>

    <!--  **************************************************************************************** -->
    <!--  Simple Types -->
    <!--  **************************************************************************************** -->
    <!-- CUSTOM TYPE DEFINITIONS-->
    <xsd:simpleType name="typeEnumType">
        <xsd:restriction base="xsd:token">
            <xsd:enumeration value="CAN"/>
            <xsd:enumeration value="CVE"/>
        </xsd:restriction>
    </xsd:simpleType>
    
    <xsd:simpleType name="statusEnumType">
        <xsd:restriction base="xsd:token">
            <xsd:enumeration value="Entry"/>
            <xsd:enumeration value="Candidate"/>
        </xsd:restriction>
    </xsd:simpleType>
    
    <!-- need to verify enumeration -->
    <xsd:simpleType name="simplePhaseEnumType">
        <xsd:restriction base="xsd:token">
            <xsd:enumeration value="Proposed"/>
            <xsd:enumeration value="Interim"/>
            <xsd:enumeration value="Modified"/>
            <xsd:enumeration value="Assigned"/>
        </xsd:restriction>
    </xsd:simpleType>

    <!--  **************************************************************************************** -->
    <!--  Complex Types -->
    <!--  **************************************************************************************** -->
    <xsd:complexType name="ItemType">
        <xsd:sequence>
            <xsd:element name="status" type="statusEnumType" minOccurs="1" maxOccurs="1"/>
            <xsd:element name="phase" type="specificPhaseType" minOccurs="0" maxOccurs="1"/>
            <xsd:element name="desc" type="xsd:string" minOccurs="1" maxOccurs="1"/>
            <xsd:element name="refs" type="refsType" minOccurs="1" maxOccurs="1"/>
            <xsd:element name="votes" type="votesType" minOccurs="0" maxOccurs="1"/>
            <xsd:element name="comments" type="commentsType" minOccurs="0" maxOccurs="1"/>
        </xsd:sequence>
        <!--Need to Verify Enumeration-->
        <xsd:attribute name="type" type="typeEnumType" use="required"/>
        <xsd:attribute name="name" type="xsd:token" use="required"/>
        <xsd:attribute name="seq" type="xsd:token" use="required"/>
    </xsd:complexType>

    <xsd:complexType name="commentsType">
        <xsd:sequence>
            <xsd:element name="comment" minOccurs="0" maxOccurs="unbounded">
                <xsd:complexType>
                    <xsd:simpleContent>
                        <xsd:extension base="xsd:string">
                            <xsd:attribute name="voter" type="xsd:token" use="required"/>
                        </xsd:extension>
                    </xsd:simpleContent>
                </xsd:complexType>
            </xsd:element>
        </xsd:sequence>
    </xsd:complexType>

    <xsd:complexType name="votesType">
        <xsd:sequence>
            <xsd:element name="accept" minOccurs="0" maxOccurs="1">
                <xsd:complexType>
                    <xsd:simpleContent>
                        <xsd:extension base="xsd:string">
                            <xsd:attribute name="count" type="xsd:token" use="required"/>
                        </xsd:extension>
                    </xsd:simpleContent>
                </xsd:complexType>
            </xsd:element>
            <xsd:element name="modify" minOccurs="0" maxOccurs="1">
                <xsd:complexType>
                    <xsd:simpleContent>
                        <xsd:extension base="xsd:string">
                            <xsd:attribute name="count" type="xsd:token" use="required"/>
                        </xsd:extension>
                    </xsd:simpleContent>
                </xsd:complexType>
            </xsd:element>
            <xsd:element name="noop" minOccurs="0" maxOccurs="1">
                <xsd:complexType>
                    <xsd:simpleContent>
                        <xsd:extension base="xsd:string">
                            <xsd:attribute name="count" type="xsd:token" use="required"/>
                        </xsd:extension>
                    </xsd:simpleContent>
                </xsd:complexType>
            </xsd:element>
            <xsd:element name="recast" minOccurs="0" maxOccurs="1">
                <xsd:complexType>
                    <xsd:simpleContent>
                        <xsd:extension base="xsd:string">
                            <xsd:attribute name="count" type="xsd:token" use="required"/>
                        </xsd:extension>
                    </xsd:simpleContent>
                </xsd:complexType>
            </xsd:element>
            <xsd:element name="reject" minOccurs="0" maxOccurs="1">
                <xsd:complexType>
                    <xsd:simpleContent>
                        <xsd:extension base="xsd:string">
                            <xsd:attribute name="count" type="xsd:token" use="required"/>
                        </xsd:extension>
                    </xsd:simpleContent>
                </xsd:complexType>
            </xsd:element>
            <xsd:element name="reviewing" minOccurs="0" maxOccurs="1">
                <xsd:complexType>
                    <xsd:simpleContent>
                        <xsd:extension base="xsd:string">
                            <xsd:attribute name="count" type="xsd:token" use="required"/>
                        </xsd:extension>
                    </xsd:simpleContent>
                </xsd:complexType>
            </xsd:element>
            <xsd:element name="revote" minOccurs="0" maxOccurs="1">
                <xsd:complexType>
                    <xsd:simpleContent>
                        <xsd:extension base="xsd:string">
                            <xsd:attribute name="count" type="xsd:token" use="required"/>
                        </xsd:extension>
                    </xsd:simpleContent>
                </xsd:complexType>
            </xsd:element>
        </xsd:sequence>
    </xsd:complexType>


    <xsd:complexType name="specificPhaseType">
        <xsd:simpleContent>
            <xsd:extension base="simplePhaseEnumType">
                <xsd:attribute name="date" type="xsd:token" use="optional"/>
            </xsd:extension>
        </xsd:simpleContent>
    </xsd:complexType>

    <xsd:complexType name="refsType">
        <xsd:annotation>
            <xsd:documentation>holds all hyperlink elements</xsd:documentation>
        </xsd:annotation>
        <xsd:sequence>
            <xsd:element name="ref" type="refType" minOccurs="0" maxOccurs="unbounded"/>
        </xsd:sequence>
    </xsd:complexType>

    <xsd:complexType name="refType">
        <xsd:annotation>
            <xsd:documentation>Holds individual hyperlink element</xsd:documentation>
        </xsd:annotation>
        <xsd:simpleContent>
            <xsd:extension base="xsd:string">
                <xsd:attribute name="source" type="xsd:token" use="required"/>
                <xsd:attribute name="url" type="xsd:anyURI" use="optional"/>
            </xsd:extension>
        </xsd:simpleContent>
    </xsd:complexType>
</xsd:schema>

nkramer commented on api-reference/beta/resources/chatAttachment.md:
```diff
@@ -0,0 +1,51 @@
+| Property	   | Type	|Description|
+|:---------------|:--------|:----------|
+|id|string| Read-only. Unique id of the attachment|
+|contentType| string | The media type of the content attachment. It can have the following values: <br><ul><li>reference: Attachment is a link to another file. Populate the contentURL with the link to the object<br></li><li>file: Raw file attachment. Populate the contenturl field with the base64 encoding of the file in data: format<br></li><li>image/: Image type with the type of the image specifi ...
```
There's html mixed in with the markdown, I don't think that's allowed.

hamenon-ms replied:
Its essentially for formatting into lists within the table. I've seen it in quite a few more places in the documentation (Eg: Planner). I can remove it, but it makes the documentation less readable hence would want to remove it only if it's absolutely necessary.

Lauragra replied:
Some HTML is okay. Note however that you have a colon followed by a comma - please fix the typos here.

hamenon-ms replied:
@Lauragra Noted. Will get those addressed

nkramer commented on api-reference/beta/resources/chatMention.md:
```diff
@@ -0,0 +1,47 @@
+## Properties
+| Property	   | Type	|Description|
+|:---------------|:--------|:----------|
+|type|string| Entity type that is associated with this message. It supports the following values<br><ul><li>user - Represents a user mention. Populate the id field with the AAD id of the user</li><li>bot - Represents a bot mention. Populate the id field with the bot id</li><li>team - Represents a team mention. Populate the id field with the group id for the team</li><li>channel - Represents a cha ...
```
more html

hamenon-ms replied:
Same as above

Lauragra replied:
Lists, br tags, etc. will convert and render correctly on the site.

nkramer commented on api-reference/beta/resources/chatreaction.md:
```diff
@@ -0,0 +1,46 @@
+## Properties
+| Property	   | Type	|Description|
+|:---------------|:--------|:----------|
+|type|string| THe type of reaction; planned values include <br><ul><li>Like - Like a message, content is blank in this case</li><li>Emoji - Emoji reaction, content is set to unicode value of the emoji</li><li>Label - content is set to the string in the label</li></ul>|
```
html

hamenon-ms replied:
Same as above

nkramer commented on api-reference/beta/resources/chatmessage.md:
```diff
@@ -2,15 +2,37 @@
 > **Important:** APIs under the /beta version in Microsoft Graph are in preview and are subject to change. Use of these APIs in production applications is not supported.
 
-Represents an individual chat message within a [chatThread](chatthread.md) entity.
+Represents an individual chat message within a [channel](channel.md) or chat entity. The message can be an root message or part of a thread which is defined by the replyToId property in the message.
```
What about the people who want to post messages, do we still have documentation for them?

hamenon-ms replied:
The current Post documentation will still work. We're still working on the more detailed post flow which will come in soon.

angelgolfer-ms commented on api-reference/beta/api/channel_message_get.md:
```diff
@@ -0,0 +1,193 @@
+
+> **Important:** APIs under the /beta version in Microsoft Graph are in preview and are subject to change. Use of these APIs in production applications is not supported.
+
+Retrieve a single root [message](../resouces/chatmmessage.md) in a [channel](../resources/channel.md) of a team.
```
Is there somewhere else that you describe what a "root" message is? 
It doesn't seem like this is a commonly understood concept - search doesn't return anything in the context of Microsoft Teams. 
How does a root message differ from a chatThread? We have defined a channel as a collection of chatThreads (and each chatThread is a collection of chatMessages)?

hamenon-ms replied:
We've moved away from the chatThread concept and are just calling them all chatMessages. Just that some chat messages have replies others don't (this is controlled by the replytoid in the message resource). 

The get channel messages api returns just the main message (and not it's replies) by default. That was what I was referring to. But I agree that I could be more explicit. I'll make the change in the upcoming commit.

angelgolfer-ms commented on api-reference/beta/api/channel_message_list.md:
```diff
@@ -0,0 +1,196 @@
+
+> **Important:** APIs under the /beta version in Microsoft Graph are in preview and are subject to change. Use of these APIs in production applications is not supported.
+
+Retrieve the list of root [messages](../resouces/chatmmessage.md) in a [channel](../resources/channel.md) of a team.
```
Similar question as GET - is there somewhere we explain the hierarchy, what's a root message vs chatThread?

hamenon-ms replied:
The get channel messages api returns just the main message (and not it's replies) by default. That was what I was referring to. But I agree that I could be more explicit. I'll make the change in the upcoming commit.

hamenon-ms commented:
@nkramer @Lauragra @angelgolfer-ms I've updated the PR based on the comments and addressed the build issues. Please do review and let me know if any more changes are needed.
